components/user-mgt/org.wso2.carbon.identity.user.pre.update.profile.action/src/main/java/org/wso2/carbon/identity/user/pre/update/profile/action/internal/execution/PreUpdateProfileRequestBuilder.java

```diff
@@ -83,13 +89,52 @@ public ActionExecutionRequest buildActionExecutionRequest(FlowContext flowContex
         UserActionContext userActionContext =
                 flowContext.getValue(UserActionContext.USER_ACTION_CONTEXT_REFERENCE_KEY, UserActionContext.class);
         PreUpdateProfileAction preUpdateProfileAction = (PreUpdateProfileAction) actionExecutionContext.getAction();
+        Event event  = getEvent(userActionContext, preUpdateProfileAction);
 
         return new ActionExecutionRequest.Builder()
                 .actionType(getSupportedActionType())
-                .event(getEvent(userActionContext, preUpdateProfileAction))
+                .event(event)
+                .allowedOperations(getAllowedOperations(event))
                 .build();
     }
 
+    private List<AllowedOperation> getAllowedOperations(Event event) throws ActionExecutionRequestBuilderException {
+
+        List<String> removeOrAddPaths = new ArrayList<>();
+        List<String> replacePaths = new ArrayList<>();
+        if (!event.getUser().getClaims().isEmpty()) {
+            removeOrAddPaths.add(USER_CLAIMS_PATH_PREFIX);
+            replacePaths.add(USER_CLAIMS_VALUE_PATH_PREFIX);
+            if (hasMultiValuedClaims(event.getUser().getClaims())) {
+                removeOrAddPaths.add(MULTI_VALUED_USER_CLAIMS_VALUE_PATH_PREFIX);
+                replacePaths.add(MULTI_VALUED_USER_CLAIMS_VALUE_PATH_PREFIX);
+            }
+        }
+
+        return Arrays.asList(
+                createAllowedOperation(Operation.ADD, removeOrAddPaths),
+                createAllowedOperation(Operation.REMOVE, removeOrAddPaths),
+                createAllowedOperation(Operation.REPLACE, replacePaths));
+    }
+
+    private boolean hasMultiValuedClaims(List<UserClaim> claims) throws ActionExecutionRequestBuilderException {
+
+        for (UserClaim claim : claims) {
+            if (isMultiValuedClaim(claim.getUri())) {
```

Comment:
Can't we avoid this redundant isMultiValued check per claim ?
Hope we can decide this by checking the claim value type

Reply:
Well if we want to check for multi valued claims IMO rather than relying on the value we should actually reply on the meta attribute for the claim